File: android/app/src/main/java/com/noop/garmin/GarminRealtime.kt
package com.noop.garmin

object GarminEpoch {
    /** FIT/Garmin epoch: 1989-12-31 00:00:00 UTC. */
    const val UNIX_OFFSET_SECONDS: Long = 631_065_600L
    fun unixSeconds(garminSeconds: Long): Long = UNIX_OFFSET_SECONDS + garminSeconds
    fun garminSeconds(unixSeconds: Long): Long? =
        (unixSeconds - UNIX_OFFSET_SECONDS).takeIf { it in 0..0xffff_ffffL }
}

sealed interface GarminRealtimeValue {
    val raw: ByteArray

    data class HeartRate(
        override val raw: ByteArray,
        val type: Int,
        val rawHeartRate: Int,
        val heartRateBpm: Int?,
        val restingHeartRate: Int,
    ) : GarminRealtimeValue

    data class Steps(override val raw: ByteArray, val steps: Int, val goal: Int) : GarminRealtimeValue

    data class Hrv(
        override val raw: ByteArray,
        val rawRrMilliseconds: Int,
        val rrMilliseconds: Int?,
        val unknown: Int,
    ) : GarminRealtimeValue

    data class Spo2(
        override val raw: ByteArray,
        val rawValue: Int,
        val percent: Int?,
        val garminTimestamp: Long,
        val unixSeconds: Long,
    ) : GarminRealtimeValue

    data class Respiration(
        override val raw: ByteArray,
        val rawValue: Int,
        val breathsPerMinute: Int?,
    ) : GarminRealtimeValue

    data class Opaque(
        val service: GarminMultiLinkService,
        override val raw: ByteArray,
    ) : GarminRealtimeValue
}

object GarminRealtimeDecoder {
    fun decode(service: GarminMultiLinkService, payload: ByteArray): GarminRealtimeValue? {
        return when (service) {
        GarminMultiLinkService.REALTIME_HEART_RATE -> {
            if (payload.size != 3 &&
                !(payload.size == 5 && payload[3] == 0xff.toByte() && payload[4] == 0xff.toByte())
            ) return null
            val rawHr = payload[1].toInt() and 0xff
            GarminRealtimeValue.HeartRate(
                raw = payload.copyOf(),
                type = payload[0].toInt() and 0xff,
                rawHeartRate = rawHr,
                heartRateBpm = rawHr.takeIf { it > 0 },
                restingHeartRate = payload[2].toInt() and 0xff,
            )
        }
        GarminMultiLinkService.REALTIME_STEPS -> {
            if (payload.size != 8) return null
            val steps = i32(payload, 0)
            val goal = i32(payload, 4)
            if (steps < 0 || goal < 0) return null
            GarminRealtimeValue.Steps(payload.copyOf(), steps, goal)
        }
        GarminMultiLinkService.REALTIME_HRV -> {
            if (payload.size != 6) return null
            val rawRr = i16(payload, 0)
            GarminRealtimeValue.Hrv(
                raw = payload.copyOf(),
                rawRrMilliseconds = rawRr,
                rrMilliseconds = rawRr.takeIf { it in 200..3_000 },
                unknown = i32(payload, 2),
            )
        }
        GarminMultiLinkService.REALTIME_SPO2 -> {
            if (payload.size != 5) return null
            val raw = payload[0].toInt()
            val timestamp = u32(payload, 1)
            GarminRealtimeValue.Spo2(
                raw = payload.copyOf(),
                rawValue = raw,
                percent = raw.takeIf { it in 1..100 },
                garminTimestamp = timestamp,
                unixSeconds = GarminEpoch.unixSeconds(timestamp),
            )
        }
        GarminMultiLinkService.REALTIME_RESPIRATION -> {
            if (payload.size != 1) return null
            val raw = payload[0].toInt()
            GarminRealtimeValue.Respiration(
                raw = payload.copyOf(),
                rawValue = raw,
                breathsPerMinute = raw.takeIf { it in 1..120 },
            )
        }
        GarminMultiLinkService.REALTIME_ACCELEROMETER,
        GarminMultiLinkService.REALTIME_BODY_BATTERY ->
            payload.takeIf { it.isNotEmpty() }?.let {
                GarminRealtimeValue.Opaque(service, it.copyOf())
            }
            GarminMultiLinkService.GFDI -> null
        }
    }

    private fun i16(bytes: ByteArray, offset: Int): Int =
        (((bytes[offset].toInt() and 0xff) or ((bytes[offset + 1].toInt() and 0xff) shl 8)).toShort()).toInt()

    private fun i32(bytes: ByteArray, offset: Int): Int =
        (bytes[offset].toInt() and 0xff) or
            ((bytes[offset + 1].toInt() and 0xff) shl 8) or
            ((bytes[offset + 2].toInt() and 0xff) shl 16) or
            ((bytes[offset + 3].toInt() and 0xff) shl 24)

    private fun u32(bytes: ByteArray, offset: Int): Long = i32(bytes, offset).toLong() and 0xffff_ffffL
}

File: Packages/GarminProtocol/Sources/GarminProtocol/GarminRealtime.swift
import Foundation

public enum GarminEpoch {
    /// FIT/Garmin epoch: 1989-12-31 00:00:00 UTC.
    public static let unixOffsetSeconds: Int64 = 631_065_600
    public static func unixSeconds(fromGarminSeconds seconds: UInt32) -> Int64 {
        unixOffsetSeconds + Int64(seconds)
    }
    public static func garminSeconds(fromUnixSeconds seconds: Int64) -> UInt32? {
        let delta = seconds - unixOffsetSeconds
        return UInt32(exactly: delta)
    }
}

public struct GarminRealtimeHeartRate: Equatable, Sendable {
    public let raw: [UInt8]; public let type: UInt8; public let rawHeartRate: UInt8
    public let heartRateBPM: Int?; public let restingHeartRate: UInt8
}
public struct GarminRealtimeSteps: Equatable, Sendable {
    public let raw: [UInt8]; public let steps: Int32; public let goal: Int32
}
public struct GarminRealtimeSpO2: Equatable, Sendable {
    public let raw: [UInt8]; public let rawValue: Int8; public let percent: Int?
    public let garminTimestamp: UInt32; public let unixSeconds: Int64
}
public struct GarminRealtimeRespiration: Equatable, Sendable {
    public let raw: [UInt8]; public let rawValue: Int8; public let breathsPerMinute: Int?
}
public struct GarminRealtimeHRV: Equatable, Sendable {
    public let raw: [UInt8]; public let rawRRMilliseconds: Int16; public let rrMilliseconds: Int?
    public let unknown: Int32
}
public struct GarminOpaqueRealtime: Equatable, Sendable {
    public let service: GarminMultiLinkService; public let raw: [UInt8]
}

public enum GarminRealtimeValue: Equatable, Sendable {
    case heartRate(GarminRealtimeHeartRate), steps(GarminRealtimeSteps), hrv(GarminRealtimeHRV)
    case spo2(GarminRealtimeSpO2), respiration(GarminRealtimeRespiration)
    case accelerometer(GarminOpaqueRealtime), bodyBattery(GarminOpaqueRealtime)
}

public enum GarminRealtimeDecoder {
    public static func decode(service: GarminMultiLinkService, payload b: [UInt8]) -> GarminRealtimeValue? {
        switch service {
        case .realtimeHeartRate:
            guard b.count == 3 || (b.count == 5 && b.suffix(2) == [0xFF, 0xFF]) else { return nil }
            let bpm = b[1] == 0 ? nil : Int(b[1])
            return .heartRate(.init(raw: b, type: b[0], rawHeartRate: b[1], heartRateBPM: bpm,
                                    restingHeartRate: b[2]))
        case .realtimeSteps:
            guard b.count == 8 else { return nil }
            let steps = i32(b, 0), goal = i32(b, 4)
            guard steps >= 0, goal >= 0 else { return nil }
            return .steps(.init(raw: b, steps: steps, goal: goal))
        case .realtimeHRV:
            guard b.count == 6 else { return nil }
            let rawRR = i16(b, 0)
            let qualified = (200...3_000).contains(Int(rawRR)) ? Int(rawRR) : nil
            return .hrv(.init(raw: b, rawRRMilliseconds: rawRR, rrMilliseconds: qualified,
                              unknown: i32(b, 2)))
        case .realtimeSpO2:
            guard b.count == 5 else { return nil }
            let raw = Int8(bitPattern: b[0]), timestamp = u32(b, 1)
            let qualified = raw > 0 && raw <= 100 ? Int(raw) : nil
            return .spo2(.init(raw: b, rawValue: raw, percent: qualified,
                               garminTimestamp: timestamp,
                               unixSeconds: GarminEpoch.unixSeconds(fromGarminSeconds: timestamp)))
        case .realtimeRespiration:
            guard b.count == 1 else { return nil }
            let raw = Int8(bitPattern: b[0])
            return .respiration(.init(raw: b, rawValue: raw,
                                      breathsPerMinute: raw > 0 && raw <= 120 ? Int(raw) : nil))
        case .realtimeAccelerometer:
            return b.isEmpty ? nil : .accelerometer(.init(service: service, raw: b))
        case .realtimeBodyBattery:
            return b.isEmpty ? nil : .bodyBattery(.init(service: service, raw: b))
        case .gfdi:
            return nil
        }
    }

    private static func u32(_ b: [UInt8], _ i: Int) -> UInt32 {
        UInt32(b[i]) | UInt32(b[i+1]) << 8 | UInt32(b[i+2]) << 16 | UInt32(b[i+3]) << 24
    }
    private static func i16(_ b: [UInt8], _ i: Int) -> Int16 {
        Int16(bitPattern: UInt16(b[i]) | UInt16(b[i+1]) << 8)
    }
    private static func i32(_ b: [UInt8], _ i: Int) -> Int32 {
        Int32(bitPattern: u32(b, i))
    }
}
